httpcore/_async/http2.py

```diff
@@ -314,17 +314,41 @@ async def send_headers(
     ) -> None:
         scheme, hostname, port, path = url
         default_port = {b"http": 80, b"https": 443}.get(scheme)
-        if port is None or port == default_port:
+
+        # In HTTP/2 the ':authority' pseudo-header is used instead of 'Host'.
+        # For convenience we automatically treat any 'Host' header value
+        # as a value for ':authority', allowing users to pass 'Host' whenever they
+        # want to customize the host header regardless of the HTTP version in use.
+        authority = None
+        for k, v in headers:
+            if k == b"host":
+                authority = v
+                break
+
+        if authority is None:
             authority = hostname
-        else:
-            authority = b"%s:%d" % (hostname, port)
+
+        if port is not None and port != default_port:
+            authority = b"%s:%d" % (authority, port)
```

Comment:
This case shouldn't override an explicit `Host` header, so I think there's a bit of tweaking needed here.

Reply:
It doesn't. At this point, authority has the value of an explicit Host header.

Reply:
Right, tho we don't want to append the port here.